sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala

```diff
@@ -373,11 +457,24 @@ case class Subtract(
 
   override def decimalMethod: String = "$minus"
 
+  override def decimalType(p1: Int, s1: Int, p2: Int, s2: Int): DecimalType = {
+    val resultScale = max(s1, s2)
+    if (allowPrecisionLoss) {
```

Comment:
It seems we can put this if-else in the base class `DecimalArithmetic`. The sub-classes should focus on calculating the original result precision and scale.

Reply:
nvm, division is very different.

Reply:
can we write the code in this way?
```
val resultPrecision = ...
val resultScale = ...
if (allowPrecisionLoss) {
  DecimalType.adjustPrecisionScale(resultPrecision, resultScale)
} else {
  ...
}
```

Reply:
rewriten